diff --git a/src/Makefile b/src/Makefile
index e582eed..30e7d2d 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1508,6 +1508,7 @@ OBJ = \
 	objects/if_cscope.o \
 	objects/if_xcmdsrv.o \
 	objects/main.o \
+	objects/vim_pepper.o \
 	objects/mark.o \
 	objects/memfile.o \
 	objects/memline.o \
@@ -1698,7 +1699,7 @@ $(VIMTARGET): auto/config.mk objects $(OBJ) version.c version.h
 		MAKE="$(MAKE)" sh $(srcdir)/link.sh
 
 xxd/xxd$(EXEEXT): xxd/xxd.c
-	cd xxd; CC="$(CC)" CFLAGS="$(CPPFLAGS) $(CFLAGS)" \
+	cd xxd; CC="$(CC)" EXEEXT="$(EXEEXT)" CFLAGS="$(CPPFLAGS) $(CFLAGS)" \
 		$(MAKE) -f Makefile
 
 # Build the language specific files if they were unpacked.
@@ -2546,6 +2547,9 @@ objects/integration.o: integration.c
 objects/main.o: main.c
 	$(CCC) -o $@ main.c
 
+objects/vim_pepper.o: vim_pepper.c
+	$(CCC) -o $@ vim_pepper.c
+
 objects/mark.o: mark.c
 	$(CCC) -o $@ mark.c
 
diff --git a/src/auto/configure b/src/auto/configure
index d09ee92..54589cd 100755
--- a/src/auto/configure
+++ b/src/auto/configure
@@ -11642,7 +11642,9 @@ _ACEOF
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking uint32_t is 32 bits" >&5
 $as_echo_n "checking uint32_t is 32 bits... " >&6; }
 if test "$cross_compiling" = yes; then :
-  as_fn_error "could not compile program using uint32_t." "$LINENO" 5
+  # This check seems to prevent any cross compiled build of vim from working.
+  # Disabling the error, as nacl's uint32_t is indeed 32-bit.
+  #as_fn_error "could not compile program using uint32_t." "$LINENO" 5
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
@@ -12344,16 +12346,16 @@ $as_echo "yes" >&6; }
     { $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
 $as_echo "no" >&6; }
   fi
-        { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether we need -D_FORTIFY_SOURCE=1" >&5
-$as_echo_n "checking whether we need -D_FORTIFY_SOURCE=1... " >&6; }
-  if test "$gccmajor" -gt "3"; then
-    CFLAGS=`echo "$CFLAGS" | sed -e 's/-Wp,-D_FORTIFY_SOURCE=.//g' -e 's/-D_FORTIFY_SOURCE=.//g' -e 's/$/ -D_FORTIFY_SOURCE=1/'`
-    { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
-$as_echo "yes" >&6; }
-  else
-    { $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
-$as_echo "no" >&6; }
-  fi
+#        { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether we need -D_FORTIFY_SOURCE=1" >&5
+#$as_echo_n "checking whether we need -D_FORTIFY_SOURCE=1... " >&6; }
+#  if test "$gccmajor" -gt "3"; then
+#    CFLAGS=`echo "$CFLAGS" | sed -e 's/-Wp,-D_FORTIFY_SOURCE=.//g' -e 's/-D_FORTIFY_SOURCE=.//g' -e 's/$/ -D_FORTIFY_SOURCE=1/'`
+#    { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
+#$as_echo "yes" >&6; }
+#  else
+#    { $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
+#$as_echo "no" >&6; }
+#  fi
 fi
 
 
diff --git a/src/main.c b/src/main.c
index 435e607..370e91a 100644
--- a/src/main.c
+++ b/src/main.c
@@ -160,7 +160,11 @@ _cdecl
 #  endif
 VimMain
 # else
+#  if defined(__native_client__)
+nacl_vim_main
+#  else
 main
+#  endif
 # endif
 (argc, argv)
     int		argc;
diff --git a/src/memfile.c b/src/memfile.c
index c6f5fdf..88a8c9f 100644
--- a/src/memfile.c
+++ b/src/memfile.c
@@ -646,7 +646,8 @@ mf_sync(mfp, flags)
 # endif
 	    /* OpenNT is strictly POSIX (Benzinger) */
 	    /* Tandem/Himalaya NSK-OSS doesn't have sync() */
-# if defined(__OPENNT) || defined(__TANDEM)
+	    /* TODO(sbc): remove once nacl_io provides sync() */
+# if defined(__OPENNT) || defined(__TANDEM) || defined(__native_client__)
 	    fflush(NULL);
 # else
 	    sync();
diff --git a/src/xxd/Makefile b/src/xxd/Makefile
index 97bbcc7..69fdaa2 100644
--- a/src/xxd/Makefile
+++ b/src/xxd/Makefile
@@ -1,7 +1,7 @@
 # The most simplistic Makefile
 
-xxd: xxd.c
-	$(CC) $(CFLAGS) $(LDFLAGS) -DUNIX -o xxd xxd.c
+xxd$(EXEEXT): xxd.c
+	$(CC) $(CFLAGS) $(LDFLAGS) -DUNIX -o xxd$(EXEEXT) xxd.c
 
 clean:
 	rm -f xxd xxd.o
